### PR TITLE
deprecate(rules-engine): actionHandlers will become protected in v13

### PR DESCRIPTION
**Pull request by matthieu-crouzet**
## Proposed change
It's not a good practice to expose a `Set`.
I added a method `registerActionHandlers` to replace the usage of `actionHandlers.add(...)`.
It will be future proof if we want to do something else when an action handler is registered.

### CODE REVIEW

File: packages/@o3r/rules-engine/src/services/runner/rules-engine.runner.service.ts [hunk 1]
```diff
@@ -119,29 +122,37 @@ export class RulesEngineRunnerService implements OnDestroy {
...
    */
   public upsertRulesets(ruleSets: Ruleset[]) {
     this.store.dispatch(setRulesetsEntities({entities: ruleSets}));
   }
 
+  /**
+   * Add action handlers in the rules engine
+   * @param actionHandlers
+   */
+  public registerActionHandlers(...actionHandlers: RulesEngineActionHandler[]) {
```
Comment (kpanot) [inline packages/@o3r/rules-engine/src/services/runner/rules-engine.runner.service.ts:152]: To be iso you should propose an `unregister`

**Review by kpanot — APPROVED** [review]

File: packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.spec.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  it('should not touch files without references', async () => {
+    expect(initialTree.readText(untouchedFile)).not.toContain('actionHandlers.add');
+    expect(initialTree.readText(untouchedFile)).not.toContain('actionHandlers.delete');
+    tree = await lastValueFrom(callRule(useRegisterActionHandlers, initialTree, {} as SchematicContext));
+    expect(tree.readText(untouchedFile)).toBe(initialTree.readText(untouchedFile));
+  });
+
+  it('should change file with a reference to actionHandlers.add', async () => {
+    expect(initialTree.readText(fileWithActionHandlerAdd)).toContain('actionHandlers.add');
```
Comment (fpaul-1A) [inline packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.spec.ts:81]: Isn't it weird to do an `expect` on the initial tree? We are just testing the code of the `beforeEach`?

File: packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import type { Rule, Tree } from '@angular-devkit/schematics';
+import { findFilesInTree } from '@o3r/schematics';
+
+/**
+ * Replace `actionHandlers.add` or `actionHandlers.delete` by `registerActionHandlers` or `unregisterActionHandlers` in file
+ * @param tree
+ */
+export const useRegisterActionHandlers: Rule = (tree: Tree) => {
+  const files = findFilesInTree(tree.root, (file) => /actionHandlers\.(add|delete)/.test(tree.readText(file)));
```
Comment (fpaul-1A) [inline packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.ts:9]: Shouldn't we be more restrictive on the regex to avoid impacting unrelated code?
```suggestion
  const files = findFilesInTree(tree.root, (file) => /\.actionHandlers\.(add|delete)/.test(tree.readText(file)));
```

**Review by fpaul-1A — APPROVED** [review]

File: packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.spec.ts (hunk 2 above)
Comment (matthieu-crouzet) [reply packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.spec.ts:81]: I added it to ensure that it is well present in the initial tree 
because I did the `find & replace` after writing the test, so indeed  `actionHandlers.add` was never present

**Review by fpaul-1A — APPROVED** [review]

File: packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.spec.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+          runInInjectionContext(m.injector, () => {
+            inject(RulesEngineRunnerService);
+            ruleEngine.actionHandlers.add(inject(PopupActionHandler));
+          });
+        })
+        // eslint-disable-next-line no-console
+        .catch(err => console.error(err));
+    `);
+    tree = await lastValueFrom(callRule(useRegisterActionHandlers, initialTree, {} as SchematicContext));
+    expect(tree.readText(fileWithActionHandlerAdd)).not.toContain('actionHandlers.add');
```
Comment (fpaul-1A) [inline packages/@o3r/rules-engine/schematics/ng-update/v11.6/use-register-action-handlers.spec.ts:42]: you don't test that it contains `registerActionHandlers` ?

**Review by kpanot — APPROVED** [review]
